### PR TITLE
Boot mode checks

### PR DESCRIPTION
**Pull request by tescande**

### CODE REVIEW

File: product.py [hunk 1]
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
...
+            # Don't propose to upgrade a BIOS installation with a UEFI installer and conversely
+            tool = PartitionTool(self.primary_disk)
+            existing_boot_mode = constants.TARGET_BOOT_MODE_LEGACY
+            boot_partnum = tool.partitionNumber(self.boot_device)
+            boot_part = tool.getPartition(boot_partnum)
+            if 'id' in boot_part and boot_part['id'] == GPTPartitionTool.ID_EFI_BOOT:
+                existing_boot_mode = constants.TARGET_BOOT_MODE_UEFI
+
+            if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI and not constants.UEFI_INSTALLER) or \
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
```
Comment (ydirson) [inline product.py]: a bit heavy to read, what about something like this (closer to XOR)
```
if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI) != constants.UEFI_INSTALLER:
```

Will also allow to get a visually clear `if` statement without a double indent (never seen that in the wild, pretty sure the code checkers would flag it)

File: product.py [hunk 2]
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
...
+            # Don't restore a BIOS backup with a UEFI installer and conversely
+            backup_is_uefi = False
+            with open(os.path.join(b.mount_point, 'etc', 'fstab'), 'r') as fstab:
+                for line in fstab:
+                    m = re.search("/boot/efi[[:space:]]+vfat", line)
+                    if m:
+                        backup_is_uefi = True
+                        break
+            if (backup_is_uefi and not constants.UEFI_INSTALLER) or \
+                (not backup_is_uefi and constants.UEFI_INSTALLER):
```
Comment (ydirson) [inline product.py]: XOR too:
```
if backup_is_uefi != constants.UEFI_INSTALLER):
```

File: product.py [hunk 3]
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
...
+                logger.log("findXenSourceBackups: ignoring, installer mode (%s) is "
+                            "mismatched with bakcup boot mode (%s)" %
```
Comment (ydirson) [inline product.py]: "does not match"? + "bakcup" typo

File: product.py [hunk 4]
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
             backup = XenServerBackup(p, b.mount_point)
             logger.log("Found a backup: %s" % (repr(backup),))
 
+            # Don't restore a BIOS backup with a UEFI installer and conversely
+            backup_is_uefi = False
+            with open(os.path.join(b.mount_point, 'etc', 'fstab'), 'r') as fstab:
+                for line in fstab:
+                    m = re.search("/boot/efi[[:space:]]+vfat", line)
```
Comment (ydirson) [inline product.py]: do we really need to check the fstype?
I suggest something like `if line.split(maxsplit=1)[0] == "/boot/efi"`

File: product.py [hunk 5]
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
...
+            boot_partnum = tool.partitionNumber(self.boot_device)
+            boot_part = tool.getPartition(boot_partnum)
+            if 'id' in boot_part and boot_part['id'] == GPTPartitionTool.ID_EFI_BOOT:
+                existing_boot_mode = constants.TARGET_BOOT_MODE_UEFI
+
+            if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI and not constants.UEFI_INSTALLER) or \
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
+                    logger.log("Cannot upgrade %s, installer mode (%s) is mismatched with existing boot mode (%s)" %
+                                (self.primary_disk, "UEFI" if constants.UEFI_INSTALLER else "legacy", existing_boot_mode))
+                    raise Exception("Boot mode mismatch")
```
Comment (ydirson) [inline product.py]: `return False` would me more direct, this descriptive string is not used

File: product.py [hunk 6]
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
         self.mount_state()
         result = True
         try:
+            # Don't propose to upgrade a BIOS installation with a UEFI installer and conversely
+            tool = PartitionTool(self.primary_disk)
+            existing_boot_mode = constants.TARGET_BOOT_MODE_LEGACY
```
Comment (ydirson) [inline product.py]: moving this to an `else` arm would be more straightforward

**Review by ydirson — CHANGES_REQUESTED** [review]
Do we really want to check at upgrade time and restore time using different criteria?  What about putting the logic in a helper function?

File: product.py (hunk 4 above)
Comment (tescande) [reply product.py]: Actually changed it to `re.search(r'^[^#].*\s+/boot/efi', line.strip())` to ignore commented lines

File: product.py (hunk 2 above)
Comment (tescande) [reply product.py]: Fixed, thanks.

File: product.py (hunk 3 above)
Comment (tescande) [reply product.py]: Fixed, thanks.

File: product.py (hunk 1 above)
Comment (tescande) [reply product.py]: Done

File: product.py (hunk 6 above)
Comment (tescande) [reply product.py]: Done

File: product.py (hunk 5 above)
Comment (tescande) [reply product.py]: Done

File: product.py (hunk 4 above)
Comment (ydirson) [reply product.py]: still, a regex seems overkill: we only need to check for the first word, and `split()` will even take care of any leading whitespace

File: product.py [hunk 7]
```diff
@@ -566,6 +566,19 @@ def findXenSourceBackups():
             backup = XenServerBackup(p, b.mount_point)
             logger.log("Found a backup: %s" % (repr(backup),))
 
+            # Don't restore a BIOS backup with a UEFI installer and conversely
+            backup_is_uefi = False
```
Comment (ydirson) [inline product.py]: nitpicking and possibly a matter of taste: that one would fit neatly in an `else` arm of the `for` loop

Comment (tescande) [reply product.py]: TIL `for: else:` loop.
Fixed, thanks

File: product.py (hunk 4 above)
Comment (tescande) [reply product.py]: As discussed, we'll stick with a `re.search()`

**Comment by tescande** [comment]
> Do we really want to check at upgrade time and restore time using different criteria? What about putting the logic in a helper function?

You're right so I ended up using the `/boot/efi` entry in `/etc/fstab` for both checks.

**Review by stormi — APPROVED** [review]

File: product.py [hunk 8]
```diff
@@ -28,6 +28,18 @@ class SettingsNotAvailable(Exception):
...
 
+def is_rootfs_uefi(mount_point):
+    try:
+        with open(os.path.join(mount_point, 'etc', 'fstab'), 'r') as fstab:
+            for line in fstab:
+                m = re.search(r'^[^#].*\s+/boot/efi\s+', line.strip())
+                if m:
+                    return True
+    except:
+        pass
```
Comment (ydirson) [inline product.py]: Unrestricted `except` is usually not a good idea (and rejected by checkers).  We should specify which errors are safe to ignore.

File: product.py [hunk 9]
```diff
@@ -28,6 +28,18 @@ class SettingsNotAvailable(Exception):
 XENSERVER_7_0_0 = Version([2, 1, 0]) # Platform version
 XENSERVER_MIN_VERSION = XENSERVER_7_0_0
 
+def is_rootfs_uefi(mount_point):
+    try:
+        with open(os.path.join(mount_point, 'etc', 'fstab'), 'r') as fstab:
+            for line in fstab:
+                m = re.search(r'^[^#].*\s+/boot/efi\s+', line.strip())
```
Comment (ydirson) [inline product.py]: is `strip()` really necessary here?

**Review by ydirson — CHANGES_REQUESTED** [review]

File: product.py (hunk 8 above)
Comment (tescande) [reply product.py]: Ok, will catch `FileNotFoundError`

File: product.py (hunk 9 above)
Comment (tescande) [reply product.py]: Ok, will do it on the regex instead

**Review by ydirson — APPROVED** [review]

**Pull request merged**